File: src/cs/Ssh/SshAlgorithms.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Collections.Generic;
using Microsoft.DevTunnels.Ssh.Algorithms;
using RandomNumberGenerator = System.Security.Cryptography.RandomNumberGenerator;

#pragma warning disable CA1034 // Nested types should not be visible
#pragma warning disable CA1724 // Nested types do not conflict with namespace names
#pragma warning disable SA1402 // File may only contain a single type
#pragma warning disable SA1649 // File name should match first type name

namespace Microsoft.DevTunnels.Ssh;

/// <summary>
/// Defines algorithms that may be included in SSH session configuration for negotiation and
/// potential use in SSH sessions.
/// </summary>
/// <remarks>
/// Some algorithms may be unavailable at runtime; those are automatically excluded from
/// negotiation even if they are included in the session configuration.
///
/// Algorithm names reference:
/// http://www.iana.org/assignments/ssh-parameters/ssh-parameters.xhtml
/// </remarks>
public static class SshAlgorithms
{
	public static class KeyExchange
	{
		public static KeyExchangeAlgorithm? None { get; } = null;

		public static KeyExchangeAlgorithm DHGroup14Sha256 { get; } = new DiffieHellman(DiffieHellman.DHGroup14Sha256, 2048, HmacAlgorithm.Sha256);
		public static KeyExchangeAlgorithm DHGroup16Sha512 { get; } = new DiffieHellman(DiffieHellman.DHGroup16Sha512, 4096, HmacAlgorithm.Sha512);

#if SSH_ENABLE_ECDH
		public static KeyExchangeAlgorithm EcdhNistp256 { get; } = new ECDiffieHellman(ECDiffieHellman.EcdhNistp256, 256, HmacAlgorithm.Sha256);
		public static KeyExchangeAlgorithm EcdhNistp384 { get; } = new ECDiffieHellman(ECDiffieHellman.EcdhNistp384, 384, HmacAlgorithm.Sha384);
		public static KeyExchangeAlgorithm EcdhNistp521 { get; } = new ECDiffieHellman(ECDiffieHellman.EcdhNistp521, 521, HmacAlgorithm.Sha512);
#endif
	}

	public static class PublicKey
	{
		public static PublicKeyAlgorithm? None { get; } = null;

		public static PublicKeyAlgorithm RsaWithSha256 { get; } = new Rsa(Rsa.RsaWithSha256, HmacAlgorithm.Sha256);
		public static PublicKeyAlgorithm RsaWithSha512 { get; } = new Rsa(Rsa.RsaWithSha512, HmacAlgorithm.Sha512);

		public static PublicKeyAlgorithm ECDsaSha2Nistp256 { get; } = new ECDsa(ECDsa.ECDsaSha2Nistp256, HmacAlgorithm.Sha256);
		public static PublicKeyAlgorithm ECDsaSha2Nistp384 { get; } = new ECDsa(ECDsa.ECDsaSha2Nistp384, HmacAlgorithm.Sha384);
		public static PublicKeyAlgorithm ECDsaSha2Nistp521 { get; } = new ECDsa(ECDsa.ECDsaSha2Nistp521, HmacAlgorithm.Sha512);
	}

	public static class Encryption
	{
		public static EncryptionAlgorithm? None { get; } = null;

		public static EncryptionAlgorithm Aes256Cbc { get; } = new EncryptionAlgorithm("aes256-cbc", "AES", CipherModeEx.CBC, 256);
		public static EncryptionAlgorithm Aes256Ctr { get; } = new EncryptionAlgorithm("aes256-ctr", "AES", CipherModeEx.CTR, 256);
#if SSH_ENABLE_AESGCM
		public static EncryptionAlgorithm Aes256Gcm { get; } = new AesGcm("aes256-gcm@openssh.com", 256);
#endif
	}

	public static class Hmac
	{
		public static HmacAlgorithm? None { get; } = null;

		public static HmacAlgorithm HmacSha256 { get; } = new HmacAlgorithm("hmac-sha2-256", HmacAlgorithm.Sha256);
		public static HmacAlgorithm HmacSha512 { get; } = new HmacAlgorithm("hmac-sha2-512", HmacAlgorithm.Sha512);

		public static HmacAlgorithm HmacSha256Etm { get; } = new HmacAlgorithm("hmac-sha2-256-etm@openssh.com", HmacAlgorithm.Sha256, encryptThenMac: true);
		public static HmacAlgorithm HmacSha512Etm { get; } = new HmacAlgorithm("hmac-sha2-512-etm@openssh.com", HmacAlgorithm.Sha512, encryptThenMac: true);
	}

	public static class Compression
	{
		public static CompressionAlgorithm? None { get; } = null;
	}

	/// <summary>
	/// Gets the default provider of cryptographically random bytes.
	/// </summary>
	public static IRandom Random { get; } = new RNG();

	private class RNG : IRandom
	{
		private readonly RandomNumberGenerator random = RandomNumberGenerator.Create();

		public void GetBytes(Buffer buffer)
		{
			this.random.GetBytes(buffer.Array, buffer.Offset, buffer.Count);
		}
	}
}

File: src/ts/ssh/algorithms/sshAlgorithms.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { SshAlgorithm } from './sshAlgorithm';
import { KeyExchangeAlgorithm, KeyExchange } from './keyExchangeAlgorithm';
import { PublicKeyAlgorithm, KeyPair, RsaParameters, ECParameters } from './publicKeyAlgorithm';
import { EncryptionAlgorithm, Cipher } from './encryptionAlgorithm';
import {
	HmacAlgorithm,
	Signer,
	Verifier,
	MessageSigner,
	MessageVerifier,
	HmacInfo,
} from './hmacAlgorithm';
import { CompressionAlgorithm } from './compressionAlgorithm';

export {
	SshAlgorithm,
	KeyExchangeAlgorithm,
	KeyExchange,
	PublicKeyAlgorithm,
	KeyPair,
	EncryptionAlgorithm,
	Cipher,
	HmacAlgorithm,
	Signer,
	Verifier,
	MessageSigner,
	MessageVerifier,
	HmacInfo,
	CompressionAlgorithm,
	RsaParameters,
	ECParameters,
};

// Swap imports to node crypto implementations when web crypto is not available.
const useWebCrypto = typeof window !== 'undefined' &&
	!!(typeof crypto === 'object' && crypto.subtle);

import { WebDiffieHellman, WebECDiffieHellman } from './web/webKeyExchange';
import { WebRsa } from './web/webRsa';
import { WebECDsa } from './web/webECDsa';
import { WebEncryption } from './web/webEncryption';
import { WebHmac } from './web/webHmac';
import { WebRandom } from './web/webRandom';

export interface Random {
	getBytes(buffer: Buffer): void;
}

/* eslint-disable @typescript-eslint/naming-convention, id-match */
const DiffieHellman: typeof WebDiffieHellman = useWebCrypto
	? WebDiffieHellman
	: require('./node/nodeKeyExchange').NodeDiffieHellman;
const ECDiffieHellman: typeof WebECDiffieHellman = useWebCrypto
	? WebECDiffieHellman
	: require('./node/nodeKeyExchange').NodeECDiffieHellman;
const Rsa: typeof WebRsa = useWebCrypto ? WebRsa : require('./node/nodeRsa').NodeRsa;
const ECDsa: typeof WebECDsa = useWebCrypto ? WebECDsa : require('./node/nodeECDsa').NodeECDsa;
const Encryption: typeof WebEncryption = useWebCrypto
	? WebEncryption
	: require('./node/nodeEncryption').NodeEncryption;
const Hmac: typeof WebHmac = useWebCrypto ? WebHmac : require('./node/nodeHmac').NodeHmac;
// eslint-disable-next-line no-redeclare
const Random: typeof WebRandom = useWebCrypto ? WebRandom : require('./node/nodeRandom').NodeRandom;
/* eslint-enable @typescript-eslint/naming-convention, id-match */

// eslint-disable-next-line no-redeclare
namespace Rsa {
	// eslint-disable-next-line no-shadow,@typescript-eslint/no-shadow
	export type KeyPair = WebRsa.KeyPair;
}

// eslint-disable-next-line no-redeclare
namespace ECDsa {
	// eslint-disable-next-line no-shadow,@typescript-eslint/no-shadow
	export type KeyPair = WebECDsa.KeyPair;
}

export { Rsa, ECDsa, Encryption };

export class SshAlgorithms {
	public static keyExchange: { [id: string]: KeyExchangeAlgorithm | null } = {
		none: null,
		dhGroup14Sha256: new DiffieHellman('diffie-hellman-group14-sha256', 2048, 'SHA2-256'),
		dhGroup16Sha512: new DiffieHellman('diffie-hellman-group16-sha512', 4096, 'SHA2-512'),
		ecdhNistp256Sha256: new ECDiffieHellman('ecdh-sha2-nistp256', 256, 'SHA2-256'),
		ecdhNistp384Sha384: new ECDiffieHellman('ecdh-sha2-nistp384', 384, 'SHA2-384'),
		ecdhNistp521Sha512: new ECDiffieHellman('ecdh-sha2-nistp521', 521, 'SHA2-512'),
	};

	public static publicKey: { [id: string]: PublicKeyAlgorithm | null } = {
		none: null,
		rsaWithSha256: new Rsa('rsa-sha2-256', 'SHA2-256'),
		rsaWithSha512: new Rsa('rsa-sha2-512', 'SHA2-512'),
		ecdsaSha2Nistp256: new ECDsa('ecdsa-sha2-nistp256', 'SHA2-256'),
		ecdsaSha2Nistp384: new ECDsa('ecdsa-sha2-nistp384', 'SHA2-384'),
		ecdsaSha2Nistp521: new ECDsa('ecdsa-sha2-nistp521', 'SHA2-512'),
	};

	public static encryption: { [id: string]: EncryptionAlgorithm | null } = {
		none: null,
		////aes256Cbc: new Encryption('aes256-cbc', 'AES', 'CBC', 256) },
		aes256Ctr: new Encryption('aes256-ctr', 'AES', 'CTR', 256),
		aes256Gcm: new Encryption('aes256-gcm@openssh.com', 'AES', 'GCM', 256),
	};

	public static hmac: { [id: string]: HmacAlgorithm | null } = {
		none: null,
		hmacSha256: new Hmac('hmac-sha2-256', 'SHA2-256'),
		hmacSha512: new Hmac('hmac-sha2-512', 'SHA2-512'),
		hmacSha256Etm: new Hmac('hmac-sha2-256-etm@openssh.com', 'SHA2-256', true),
		hmacSha512Etm: new Hmac('hmac-sha2-512-etm@openssh.com', 'SHA2-512', true),
	};

	public static compression: { [id: string]: CompressionAlgorithm | null } = {
		none: null,
	};

	public static random: Random = new Random();
}

export function algorithmNames<T extends SshAlgorithm>(list: (T | null)[]): string[] {
	return list.map((a) => (a ? a.name : 'none'));
}
